### PR TITLE
Bump Jaeger version to 1.14.0 (latest)

### LINKED ISSUE #97: Upgrade Jaeger to latest version
Jaeger version that we use has this bug: https://github.com/jaegertracing/jaeger/pull/1625

The bug is already fixed, we need to update to the latest version and rollback this workaround: https://github.com/open-telemetry/opentelemetry-service/pull/96

Upgrading to latest Jaeger is not straightforward. They have breaking changes.

### PR DESCRIPTION
**Pull request by annanay25**
Signed-off-by: Annanay <cs14btech11001@iith.ac.in>

### Objective

- Bump Jaeger version to 1.14.0 (latest) - Fixes #97 

### Changes

- Refactor healthcheck and jaegerreceiver modules

cc @yurishkuro @jpkrohling

### CODE REVIEW

**Comment by annanay25** [comment]
I signed it

File: extension/healthcheckextension/healthcheckextension.go [hunk 1]
```diff
@@ -35,9 +37,20 @@ var _ (extension.PipelineWatcher) = (*healthCheckExtension)(nil)
 func (hc *healthCheckExtension) Start(host extension.Host) error {
 
 	hc.logger.Info("Starting health-check extension", zap.Any("config", hc.config))
+
```
Comment (tigrannajaryan) [inline extension/healthcheckextension/healthcheckextension.go]: Why is the healthCheckExtension changed in a PR that is about Bump Jaeger version? Are these 2 things related in any way? If not please split, we want one PR per logical change.

Comment (annanay25) [reply extension/healthcheckextension/healthcheckextension.go]: The `healthcheck` package in jaeger has changed since v1.9.0 (http.Server was moved out of Healthcheck struct). These changes in `healthcheckextension` fix the following compile time errors -

```
# github.com/open-telemetry/opentelemetry-service/extension/healthcheckextension
extension/healthcheckextension/healthcheckextension.go:40:25: hc.server.Serve undefined (type *healthcheck.HealthCheck has no field or method Serve)
extension/healthcheckextension/healthcheckextension.go:66:26: too many arguments in call to healthcheck.New
extension/healthcheckextension/healthcheckextension.go:66:52: undefined: healthcheck.Logger
# github.com/open-telemetry/opentelemetry-service/extension/healthcheckextension [github.com/open-telemetry/opentelemetry-service/extension/healthcheckextension.test]
extension/healthcheckextension/healthcheckextension.go:40:25: hc.server.Serve undefined (type *healthcheck.HealthCheck has no field or method Serve)
extension/healthcheckextension/healthcheckextension.go:66:26: too many arguments in call to healthcheck.New
extension/healthcheckextension/healthcheckextension.go:66:52: undefined: healthcheck.Logger
```

Comment (tigrannajaryan) [reply extension/healthcheckextension/healthcheckextension.go]: @pjanotti can you please chime in here? I wasn't aware that healtcheck extension had a direct interaction with Jaeger package. Was this intended?

Comment (pjanotti) [reply extension/healthcheckextension/healthcheckextension.go]: @tigrannajaryan actually this a dependency that we want to remove but we have to refactor it to our own implementation, see https://github.com/open-telemetry/opentelemetry-service/issues/307

Comment (annanay25) [reply extension/healthcheckextension/healthcheckextension.go]: Aha, my bad, did not see that issue. Will first work on refactoring.

Comment (tigrannajaryan) [reply extension/healthcheckextension/healthcheckextension.go]: @annanay25 I think it is OK for you to merge this PR as is since it is valuable on its own and then we can do the refactoring (feel free to do it yourself if you want). @pjanotti are you OK with this approach?

Comment (pjanotti) [reply extension/healthcheckextension/healthcheckextension.go]: Yes, @tigrannajaryan and @annanay25 let's merge as it is and do the refactor in a separate PR.

Comment (annanay25) [reply extension/healthcheckextension/healthcheckextension.go]: Thanks!

**Comment by annanay25** [comment]
@pjanotti - I've fixed all the healthcheck tests, not sure why `TestApplication_StartUnified` is failing - just checking if I'm missing something obvious. I'll continue debugging tomorrow.

**Comment by tigrannajaryan** [comment]
Reviews, please have a look.

File: receiver/jaegerreceiver/trace_receiver.go [hunk 2]
```diff
@@ -278,6 +289,30 @@ func (jr *jReceiver) SubmitBatches(ctx thrift.Context, batches []*jaeger.Batch)
 	return jbsr, nil
 }
 
+func (jtr *jTchannelReceiver) SubmitBatches(ctx thrift.Context, batches []*jaeger.Batch) ([]*jaeger.BatchSubmitResponse, error) {
```
Comment (owais) [inline receiver/jaegerreceiver/trace_receiver.go]: Can this and the other SubmitBatches function may be share call a shared internal func instead of duplicating?

Comment (pjanotti) [reply receiver/jaegerreceiver/trace_receiver.go]: Just to confirm: before that same signature satisfied two interfaces (`JaegerBatchesHandler`and `TChanCollector`) in version 1.14 those diverged, correct?

**Comment by annanay25** [comment]
Thanks for the reviews @pjanotti, @owais - I've addressed the comments.

File: receiver/jaegerreceiver/trace_receiver.go (hunk 2 above)
Comment (annanay25) [reply receiver/jaegerreceiver/trace_receiver.go]: Yes, precisely @pjanotti

File: extension/healthcheckextension/healthcheckextension.go [hunk 3]
```diff
@@ -35,9 +38,21 @@ var _ (extension.PipelineWatcher) = (*healthCheckExtension)(nil)
...
+	// Initialize listener
+	portStr := ":" + strconv.Itoa(int(hc.config.Port))
```
[the remark this answers is not in the document]
Comment (annanay25) [reply extension/healthcheckextension/healthcheckextension.go]: Sure, will take up the refactor in another PR to add a healthcheck package in OTel. Meanwhile I felt this update could unblock a bunch of other work as well

File: receiver/jaegerreceiver/trace_receiver.go [hunk 4]
```diff
@@ -253,10 +262,10 @@ func (jr *jReceiver) stopTraceReceptionLocked() error {
 }
 
 const collectorReceiverTagValue = "jaeger-collector"
+const tchannelCollectorReceiverTagValue = "jaeger-tchannel-collector"
 
-func (jr *jReceiver) SubmitBatches(ctx thrift.Context, batches []*jaeger.Batch) ([]*jaeger.BatchSubmitResponse, error) {
+func consumeTraceData(ctx context.Context, batches []*jaeger.Batch, consumer consumer.TraceConsumer) ([]*jaeger.BatchSubmitResponse, error) {
```
Comment (owais) [inline receiver/jaegerreceiver/trace_receiver.go]: A stab in the dark but are there any implications of passing an interface value as an argument on hot path? Will accepting an interface value and calling a method on it incur some perf penalty significant enough for the hot path? I don't know if it is something to worry about but wanted to point it out in case someone has an opinion. Assuming it does, would resolving the method to a function type and then passing a callback be any better at all?

Comment (annanay25) [reply receiver/jaegerreceiver/trace_receiver.go]: Hi @owais, thanks for pointing that out.
However I'm not sure I'm clear about the fix? If I understand correctly a dynamic dispatch for the interface will be made regardless of this change? Please correct me if I'm wrong here.

Comment (owais) [reply receiver/jaegerreceiver/trace_receiver.go]: I think you are right. Just something that caught my eye and I wasn't sure about it. I was thinking that having the object on a struct would resolve the "real" type only once but I suppose that won't happen anyway unless we convert  the interface value to a concrete type so it's all probably the same. Outside the scope of this PR, it would be interesting to benchmark interface vs concrete type method calls on the hot path and see if it is even significant enough.

Comment (pjanotti) [reply receiver/jaegerreceiver/trace_receiver.go]: > it would be interesting to benchmark interface vs concrete type method calls on the hot path and see if it is even significant enough.

Yeah, that sounds a reasonable approach here. What occurred to me was to pass the function to be called directly.

**Comment by annanay25** [comment]
Thanks!
